orderer/common/multichannel/registrar.go

```diff
@@ -430,6 +430,11 @@ func (r *Registrar) JoinChannel(channelID string, configBlock *cb.Block) (types.
 		return types.ChannelInfo{}, types.ErrChannelAlreadyExists
 	}
 
+	_, _, err := r.validateJoinBlock(channelID, configBlock)
```

Comment:
we are coupling here validation and future implementation of joining a block.

Why is this here in the so called "back end" and not closer to the request processing?

Reply:
The "front-end" in the channelparticipation  package has no knowledge of the channels that exist. We therefore want to check that the operation is permitted before looking deeper into the block. The boundary we decided on is that the front-end just unmarshals the []byte into a *common.Block, and the "back-end" looks into the block.
The next PR will show how we use the result of this one. You can look at a draft here:
https://github.com/tock-ibm/fabric/tree/draft-join-member-no-onboarding

Reply:
> We therefore want to check that the operation is permitted before looking deeper into the block.

I don't think processing time optimization is a justification to couple together block validation and channel mapping. 

You can not check if the channel already exists and then return that it exists in the registrar, but in my opinion block validation should be checked closer to the input processing. 

In general I am not in favor of doing validation in the backend because it makes testing more complex. If you have a layer that only does validation checks then you can isolate it nicely and test it easier.

Reply:
I don't necessarily think it is a time optimization, but rather a division of work, and the logical flow of processing.  We can do some of the validation that is consenter agnostic and non-state related in the front-end, for sure. For example, checking the consistency of the channelID in the block with the url, and such. 

However, other parts of validation, those that are related to the state and are specific to a consenter should be done later, in the Registrar. The Registrar actually delegates part of the validation to the consenter, to keep the consenter specific (raft, in this case) logic close together, and because the registrar cannot call etcdraft functions because of an import cycle.

BTW, the import cycle can be broken quite easily, but still, I think the style of this code is  to prefer that the multichannel package will not depend on a specific consensus package but use the indirection of the Consenter abstraction.

Reply:
Yes, I understand that the flow goes like this:
**channelParticipationAPI --> registrar --> consenter**

But if you observe, you don't really need the consenter at all. You're using a method of the consenter without a real need for it and you can replace it with **ConsenterCertificate** in the etcdraft package like I hinted in the CR. 

So, I am saying that going through the registrar and then into the consenter is un-needed and instead you can just create an object which block validation and wire it into the **channelParticipationAPI** at its creation time, according to the consenter type. 

You can do it in 2 ways: 
1) Have another interface that just validates join requests 
2) Have a proxy struct which embeds the [ChannelManagement](https://github.com/hyperledger/fabric/blob/master/orderer/common/channelparticipation/restapi.go#L41) and implements only the **JoinChannel** and adds validation checks to it, and then calls further into the next **ChannelManagement** (which is really the registrar) so you'll have:
**channelManagementAPI --> validation layer ---> registrar** and this way you can keep the registrar doing what it is supposed to do and not put additional responsibility to it.

Reply:
The way to the ConsenterCertificate goes through the consenter, because it is the one initialized with all this material. The way to the consenter goes through the registrar. So i don't see any value in complicating the flow:

api -> abstract-validate(really regeistrar->consenter-validate) -> api -> join -> registrar -> consenter -> handle-chain

instead of 

 api -> join -> registrar ->consenter-validate -> consenter->handle-chain

Note that the type of checks performed in the first part of the current validate method (i.e. non-consenter specific) is not something foreign to the registrar... it does essentially the same [https://github.com/tock-ibm/fabric/blob/73e5a537deeacc55dff3b352b478a4ae4e45a625/orderer/common/multichannel/registrar.go#L284](here, checking resources), and [https://github.com/tock-ibm/fabric/blob/73e5a537deeacc55dff3b352b478a4ae4e45a625/orderer/common/multichannel/registrar.go#L129](parsing config)...

Reply:
> The way to the ConsenterCertificate goes through the consenter, because it is the one initialized with all this material. 

It has the TLS certificate in the consenter, true - but you can also acquire it in the same place where the ParticipationAPI is initialized (the main). 

> 
>  The way to the consenter goes through the registrar. So i don't see any value in complicating the flow:
> 
> api -> abstract-validate(really regeistrar->consenter-validate) -> api -> join -> registrar -> consenter -> handle-chain
> 
> instead of
> 
> api -> join -> registrar ->consenter-validate -> consenter->handle-chain

That's not exactly what I proposed.

I proposed to move all validation logic to a separate struct which would only do that, and not couple the consenter object to validation at all. 
If you need something from the consensus/etcdraft package you can inject a function/interface from where that validation object will be initialized (the same place as the ParticipationAPI will).


> Note that the type of checks performed in the first part of the current validate method (i.e. non-consenter specific) is not something foreign to the registrar... it does essentially the same here, checking resources, and here parsing config ...

This is different because the resource checking is done to initialize a bundle so all errors along the way are fatal errors. 
Similar thing for obtaining the last config from the ledger - it's not a **check**, it is **parsing** without recovery resolution.

In contrast, what you have in your case is validation checks which I do not believe are the responsibility of the registrar.